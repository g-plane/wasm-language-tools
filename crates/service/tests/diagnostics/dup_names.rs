use super::*;
use insta::assert_json_snapshot;
use lsp_types::Uri;
use wat_service::LanguageService;

#[test]
fn same_kind() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "
(module
  (func $f (param i32))
  (func $f (result f32)))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    allow_unused(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn different_kinds() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "
(module
  (global $f i32)
  (func $f))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    allow_unused(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn param_and_local() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "
(module
  (func (param $a i32) (local $a f32)))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    allow_unused(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn different_scopes() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "
(module
  (func (param $p i32))
  (func (param $p i32)))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    allow_unused(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert!(pick_diagnostics(response).is_empty());
}

#[test]
fn ref_idx() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "
(module
  (func $fn
    (call $fn)
    (call $fn)))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    allow_unused(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert!(pick_diagnostics(response).is_empty());
}

#[test]
fn blocks() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = "
(module
  (func
    (block $b
      (block $a
        (block $b
          (block $c
            (block $b)))))
    (block
      (block $b))))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    allow_unused(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn exports() {
    let uri = "untitled:test".parse::<Uri>().unwrap();
    let source = r#"
(module
  (func (export "func"))
  (export "func" (func 0)))
"#;
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    allow_unused(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}
