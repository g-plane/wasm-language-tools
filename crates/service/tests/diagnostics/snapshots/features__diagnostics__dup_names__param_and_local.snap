---
source: crates/service/tests/diagnostics/dup_names.rs
expression: response
---
{
  "kind": "full",
  "items": [
    {
      "range": {
        "start": {
          "line": 2,
          "character": 15
        },
        "end": {
          "line": 2,
          "character": 17
        }
      },
      "severity": 1,
      "code": "duplicated-names",
      "source": "wat",
      "message": "duplicated param or local name `$a` in this scope",
      "relatedInformation": [
        {
          "location": {
            "uri": "untitled:test",
            "range": {
              "start": {
                "line": 2,
                "character": 30
              },
              "end": {
                "line": 2,
                "character": 32
              }
            }
          },
          "message": "already defined here as `$a`"
        }
      ]
    },
    {
      "range": {
        "start": {
          "line": 2,
          "character": 30
        },
        "end": {
          "line": 2,
          "character": 32
        }
      },
      "severity": 1,
      "code": "duplicated-names",
      "source": "wat",
      "message": "duplicated param or local name `$a` in this scope",
      "relatedInformation": [
        {
          "location": {
            "uri": "untitled:test",
            "range": {
              "start": {
                "line": 2,
                "character": 15
              },
              "end": {
                "line": 2,
                "character": 17
              }
            }
          },
          "message": "already defined here as `$a`"
        }
      ]
    }
  ]
}
